website/docs/topics/code-execution/user-defined-functions.ipynb

```diff
@@ -0,0 +1,431 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# User Defined Functions\n",
+    "\n",
+    "```mdx\n",
+    ":::note\n",
+    "This is experimental and not *yet* supported by all executors. At this stage only [`LocalCommandLineCodeExecutor`](/docs/reference/coding/local_commandline_code_executor#localcommandlinecodeexecutor) is supported.\n",
+    "\n",
+    "\n",
+    "Currently, the method of registering tools and using this feature are different. We would like to unify them. See Github issue [here](https://github.com/microsoft/autogen/issues/2101)\n",
+    ":::\n",
+    "```\n",
+    "\n",
+    "User defined functions allow you to define Python functions in your AutoGen program and then provide these to be used by your executor. This allows you to provide your agents with tools without using traditional tool calling APIs. Currently only Python is supported for this feature.\n",
+    "\n",
+    "There are several steps involved:\n",
+    "\n",
+    "1. Define the function\n",
+    "2. Provide the function to the executor\n",
+    "3. Explain to the code writing agent how to use the function\n",
+    "\n",
+    "\n",
+    "## Define the function\n",
+    "\n",
+    "```mdx\n",
+    ":::warning\n",
+    "Keep in mind that the entire source code of these functions will be available to the executor. This means that you should not include any sensitive information in the function as an LLM agent may be able to access it.\n",
+    ":::\n",
+    "```\n",
+    "\n",
+    "If the function does not require any external imports or dependencies then you can simply use the function. For example:\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def add_two_numbers(a: int, b: int) -> int:\n",
+    "    \"\"\"Add two numbers together.\"\"\"\n",
+    "    return a + b"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This would be a valid standalone function.\n",
+    "\n",
+    "```mdx\n",
+    ":::tip\n",
+    "Using type hints and docstrings are not required but are highly recommended. They will help the code writing agent understand the function and how to use it.\n",
+    ":::\n",
+    "```\n",
+    "\n",
+    "If the function requires external imports or dependencies then you can use the `@with_requirements` decorator to specify the requirements. For example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas\n",
+    "from autogen.coding.func_with_reqs import with_requirements\n",
+    "\n",
+    "\n",
+    "@with_requirements(python_packages=[\"pandas\"], global_imports=[\"pandas\"])\n",
```

Comment:
Does `global_imports` allow `from foo.bar. import pi`? If so, how do I support that? Or is that not allowed? Where is the documentation for that?

Reply:
It's definitely possible, I cant remember if documented. I need to check